runtime/vm/FastJNI_com_ibm_crypto_plus_provider_icc_NativeInterface.cpp

```diff
@@ -0,0 +1,415 @@
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "fastJNI.h"
+
+#include "j9protos.h"
+#include "j9consts.h"
+#include "objhelp.h"
+#include "BytecodeAction.hpp"
+#include "VMHelpers.hpp"
+#include "ArrayCopyHelpers.hpp"
+
+//#define CRYPTO_DEBUG 0
+
+extern "C" {
+
+typedef int CIPHER_t(long iccContextId, long iccCipherId, void* input, int inputLen, void* output, int needsReinit);
+typedef int z_kmc_native_t(void* input, void* output, int inputLength, long paramPointer, int mode);
+
+typedef union {
+	CIPHER_t* CIPHER_op;
+	z_kmc_native_t* z_kmc_native;
+} cryptoFunc_t;
+
+typedef struct cryptoPointers {
+	cryptoFunc_t CIPHER_encryptFinalPtr;
+	cryptoFunc_t CIPHER_decryptFinalPtr;
+	cryptoFunc_t CIPHER_encryptUpdatePtr;
+	cryptoFunc_t CIPHER_decryptUpdatePtr;
+	cryptoFunc_t z_kmc_nativePtr;
+	
+	CIPHER_t* CIPHER_encryptUpdate;
+	CIPHER_t* CIPHER_decryptUpdate;
+} cryptoPointers_t;
+
+
+/* loads crypto lib routines and store routine pointers for use.
+ * Need to synchronize this operation so that threads won't
+ * use the routines too early
+ * */
+void loadCryptoLib(J9VMThread *currentThread) {
+
+	if (currentThread->javaVM->jniCryptoFunctions == NULL) {
+
+		omrthread_monitor_t mutex = currentThread->javaVM->jniCryptoLibLock;
+		omrthread_monitor_enter(mutex);
+		
+		if (currentThread->javaVM->jniCryptoFunctions == NULL) {
+	
+			PORT_ACCESS_FROM_VMC(currentThread);
+			cryptoPointers_t* cryptFuncs = (cryptoPointers_t*)j9mem_allocate_memory(sizeof(cryptoPointers_t), J9MEM_CATEGORY_VM_JCL);
+			UDATA* cryptoLibrary = &(currentThread->javaVM->jniCryptoLibrary);
+	
+			CIPHER_t* CIPHER_encryptFinal;
+			CIPHER_t* CIPHER_decryptFinal;
+			z_kmc_native_t* z_kmc_native;
+			const char * cryptolib_name = "jgskit";
+		
+			const char * encryptFinal_name = "CIPHER_encryptFinal_internal";
+			const char * decryptFinal_name = "CIPHER_decryptFinal_internal";
+			const char * encryptUpdate_name = "CIPHER_encryptUpdate_internal";
+			const char * decryptUpdate_name = "CIPHER_decryptUpdate_internal";
+			const char * zKMC_name = "CIPHER_zKMC_internal";
+	
+			j9sl_open_shared_library((char*)cryptolib_name, cryptoLibrary, J9PORT_SLOPEN_DECORATE);
```

Comment:
Error checks required for all port library calls.

Reply:
updated the error handling as per feedback.